Source/Core/AudioCommon/Mixer.cpp

```diff
@@ -51,134 +105,146 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples, boo
-		m_frac &= 0xffff;
+		samples[current_sample + 1] += l_volume * l_output;
+		samples[current_sample] += r_volume * r_output;
+
+		m_fraction += ratio;
+		read_index += 2 * (s32) m_fraction;
+		m_fraction = m_fraction - (s32) m_fraction;
 	}
 
-	// Padding
-	short s[2];
-	s[0] = Common::swap16(m_buffer[(indexR - 1) & INDEX_MASK]);
-	s[1] = Common::swap16(m_buffer[(indexR - 2) & INDEX_MASK]);
-	s[0] = (s[0] * rvolume) >> 8;
-	s[1] = (s[1] * lvolume) >> 8;
-	for (; currentSample < numSamples * 2; currentSample += 2)
+	// pad output if not enough input samples
+	float s[2];
+	s[0] = m_float_buffer[(read_index - 1) & INDEX_MASK] * r_volume;
+	s[1] = m_float_buffer[(read_index - 2) & INDEX_MASK] * l_volume;
+	for (; current_sample < numSamples * 2; current_sample += 2)
 	{
-		int sampleR = s[0] + samples[currentSample];
-		MathUtil::Clamp(&sampleR, -32767, 32767);
-		samples[currentSample] = sampleR;
-		int sampleL = s[1] + samples[currentSample + 1];
-		MathUtil::Clamp(&sampleL, -32767, 32767);
-		samples[currentSample + 1] = sampleL;
+		samples[current_sample] += s[0];
+		samples[current_sample + 1] += s[1];
 	}
 
-	// Flush cached variable
-	Common::AtomicStore(m_indexR, indexR);
+	// update read index
+	Common::AtomicStore(m_read_index, read_index);
+}
 
-	return numSamples;
+// we NEED dithering going from float -> 16bit
+void CMixer::TriangleDither(float* l_sample, float* r_sample)
+{
+	float left_dither = DITHER_NOISE;
+	float right_dither = DITHER_NOISE;
+	*l_sample = (*l_sample) + left_dither - m_l_dither_prev;
+	*r_sample = (*r_sample) + right_dither - m_r_dither_prev;
+	m_l_dither_prev = left_dither;
+	m_r_dither_prev = right_dither;
 }
 
-unsigned int CMixer::Mix(short* samples, unsigned int num_samples, bool consider_framelimit)
+u32 CMixer::Mix(s16* samples, u32 num_samples, bool consider_framelimit)
 {
 	if (!samples)
 		return 0;
 
-	std::lock_guard<std::mutex> lk(m_csMixing);
-
-	memset(samples, 0, num_samples * 2 * sizeof(short));
+	std::lock_guard<std::mutex> lk(m_cs_mixing);
 
 	if (PowerPC::GetState() != PowerPC::CPU_RUNNING)
 	{
 		// Silence
+		memset(samples, 0, num_samples * 2 * sizeof(s16));
 		return num_samples;
 	}
 
-	m_dma_mixer.Mix(samples, num_samples, consider_framelimit);
-	m_streaming_mixer.Mix(samples, num_samples, consider_framelimit);
-	m_wiimote_speaker_mixer.Mix(samples, num_samples, consider_framelimit);
+	// reset float output buffer
+	m_output_buffer.resize(num_samples * 2);
+	std::fill_n(m_output_buffer.begin(), num_samples * 2, 0.f);
+
+	m_dma_mixer.Mix(m_output_buffer, num_samples, consider_framelimit);
+	m_streaming_mixer.Mix(m_output_buffer, num_samples, consider_framelimit);
+	m_wiimote_speaker_mixer.Mix(m_output_buffer, num_samples, consider_framelimit);
+
+	// dither and clamp
+	for (u32 i = 0; i < num_samples * 2; i += 2)
+	{
+		float l_output = m_output_buffer[i + 1];
+		float r_output = m_output_buffer[i];
+		TriangleDither(&m_output_buffer[i + 1], &m_output_buffer[i]);
+
```

Comment:
You are not really applying dither here as  the output for dithering is the buffer and you have already readed the values from the variables. also dithering should be aplied to the scaled variables so first you have to multiply the sanple  by 32768 then apply dithereng, clamp  and finally cast to s16.
you can see that applied here:
http://audacity.googlecode.com/svn/audacity-src/trunk/src/Dither.cpp